File: src/counter.cpp
#include "rclcpp/rclcpp.hpp"
#include "turtle_control/msg/message_counter.hpp"

using namespace std::chrono_literals;

class Counter : public rclcpp::Node
{
    public:
        Counter() : Node("message_counter")
        {
            publisher_ = this->create_publisher<turtle_control::msg::MessageCounter>("count_number", 10);

            msg_counter.count = 0;
            msg_counter.state = false;
            msg_counter.name = "Robot";

            auto pub_count = [this]() -> void
            {
                if(msg_counter.count == 10)
                {
                    msg_counter.count = 0;
                    msg_counter.state = !msg_counter.state;
                    std::cout << "Hey " << msg_counter.name << "! Starting again." << std::endl;
                    std::cout << "State changed." << std::endl;
                }

                std::cout << "Counter is " << msg_counter.count << std::endl;
                std::cout << "State is " << std::boolalpha << msg_counter.state << std::endl;

                this->publisher_->publish(msg_counter);
                msg_counter.count++;
            };

            timer_ = this->create_wall_timer(1s, pub_count);
        }
    private:
        rclcpp::Publisher<turtle_control::msg::MessageCounter>::SharedPtr publisher_;
        rclcpp::TimerBase::SharedPtr timer_;
        turtle_control::msg::MessageCounter msg_counter;
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv);
    rclcpp::spin(std::make_shared<Counter>());
    rclcpp::shutdown();

    return 0;
}

File: scripts/counter.py
#!/usr/bin/env python3

import rclpy
from rclpy.node import Node
from turtle_control.msg import MessageCounter

class Counter(Node):
    msg_counter = MessageCounter()
    msg_counter.name = "Robot"
    msg_counter.count = 0
    msg_counter.state = True

    def __init__(self):
        super().__init__("message_counter")
        self.publisher_ = self.create_publisher(MessageCounter, "count_number", 10)
        self.timer_ = self.create_timer(1.0, self.pub_count)

    def pub_count(self):
        if self.msg_counter.count == 10:
            self.msg_counter.count = 0
            self.msg_counter.state = not self.msg_counter.state
            print("Hey " + str(self.msg_counter.name) + "! Starting again.")
            print("State changed")

        print("Counter is" + str(self.msg_counter.count))
        print("State is " + str(self.msg_counter.state))

        self.publisher_.publish(self.msg_counter)
        self.msg_counter.count += 1

def main(args=None):
    rclpy.init(args=args)
    counter = Counter()
    rclpy.spin(counter)
    rclpy.shutdown()

if __name__ == "__main__":
    main()